python/tskit/trees.py

```diff
@@ -9291,25 +9290,156 @@ def ibd_segments(
+            remainder = sequence_length - position
+
+            while b < num_edges and remove_position[b] == position:  # edges out
+                e = remove_index[b]
+                p = edges_parent[e]
+                c = edges_child[e]
+                nodes_parent[c] = tskit.NULL
+                inside = sample_counts[c]
+                while p != tskit.NULL:
+                    outside = sample_counts[p] - sample_counts[c]
+                    for i, (j, k) in enumerate(indexes):
+                        weight = inside[j] * outside[k] + inside[k] * outside[j]
+                        coalescing_pairs[w, p, i] -= weight * remainder
+                    c, p = p, nodes_parent[p]
+                p = edges_parent[e]
+                while p != tskit.NULL:
+                    sample_counts[p] -= inside
+                    p = nodes_parent[p]
+                b += 1
+
+            while a < num_edges and insert_position[a] == position:  # edges in
+                e = insert_index[a]
+                p = edges_parent[e]
+                c = edges_child[e]
+                nodes_parent[c] = p
+                inside = sample_counts[c]
+                while p != tskit.NULL:
+                    sample_counts[p] += inside
+                    p = nodes_parent[p]
+                p = edges_parent[e]
+                while p != tskit.NULL:
+                    outside = sample_counts[p] - sample_counts[c]
+                    for i, (j, k) in enumerate(indexes):
+                        weight = inside[j] * outside[k] + inside[k] * outside[j]
+                        coalescing_pairs[w, p, i] += weight * remainder
+                    c, p = p, nodes_parent[p]
+                a += 1
+
+            position = sequence_length
+            if b < num_edges:
+                position = min(position, remove_position[b])
+            if a < num_edges:
+                position = min(position, insert_position[a])
+
+            while w < num_windows and windows[w + 1] <= position:  # flush window
+                remainder = sequence_length - windows[w + 1]
+                for c, p in enumerate(nodes_parent):
+                    if p == tskit.NULL:
+                        continue
+                    inside = sample_counts[c]
+                    outside = sample_counts[p] - sample_counts[c]
+                    for i, (j, k) in enumerate(indexes):
+                        weight = inside[j] * outside[k] + inside[k] * outside[j]
+                        coalescing_pairs[w, p, i] -= weight * remainder / 2
+                        if w + 1 < num_windows:
+                            coalescing_pairs[w + 1, p, i] += weight * remainder / 2
+                w += 1
+
+        for i, (j, k) in enumerate(indexes):
+            coalescing_pairs[:, :, i] /= 1 + int(j == k)
+
+        if span_normalise:
+            for w, s in enumerate(np.diff(windows)):
+                coalescing_pairs[w] /= s
+
+        return coalescing_pairs.squeeze()
```

Comment:
I think we only want to squeeze dimensions for which the corresponding argument was `None`? We could use the method for this over in `stats.py`?